include/fiction/algorithms/simulation/sidb/ground_state_space.hpp

```diff
@@ -0,0 +1,802 @@
+    get_initial_clustering(const sidb_cluster_ptr& c,
+                           const std::pair<charge_distribution_surface<Lyt>, charge_distribution_surface<Lyt>>&
+                               local_potential_bound_containers) noexcept
+    {
+        const auto& [min_loc_pot_cds, max_loc_pot_cds] = local_potential_bound_containers;
+
+        sidb_clustering clustering{};
+
+        if (c->size() == 1)
+        {
+            const uint64_t i = get_singleton_sidb_ix(c);
+
+            c->initialize_singleton_cluster_charge_space(-min_loc_pot_cds.get_local_potential_by_index(i).value(),
+                                                         -max_loc_pot_cds.get_local_potential_by_index(i).value(),
+                                                         min_loc_pot_cds.get_phys_params().base);
+
+            for (uint64_t j = 0; j < min_loc_pot_cds.num_cells(); ++j)
+            {
+                c->pot_projs[j] = potential_projection_order{min_loc_pot_cds.get_chargeless_potential_by_indices(i, j),
+                                                             min_loc_pot_cds.get_phys_params().base};
+            }
+
+            clustering.emplace(c);
+        }
+
+        for (const sidb_cluster_ptr& child : c->children)
+        {
+            const sidb_clustering& child_clustering = get_initial_clustering(child, local_potential_bound_containers);
+            clustering.insert(child_clustering.cbegin(), child_clustering.cend());
+        }
+
+        return clustering;
+    }
+
+    template <bound_direction bound>
+    static constexpr inline potential_projection get_proj_bound(const sidb_cluster_ptr& c,
+                                                                const uint64_t          sidb_ix) noexcept
+    {
+        return c->pot_projs.at(sidb_ix).get<bound>();
+    }
+
+    template <bound_direction bound>
+    static constexpr inline double get_next_proj_pot_bound(const sidb_cluster_ptr& c, const uint64_t sidb_ix) noexcept
+    {
+        return c->pot_projs.at(sidb_ix).get_next<bound>().v;
+    }
+
+    template <bound_direction bound>
+    static constexpr inline potential_projection get_proj_state_bound(const sidb_cluster_projector_state& pst,
+                                                                      const uint64_t sidb_ix) noexcept
+    {
+        return pst.cluster->pot_projs.at(sidb_ix).get_pot_proj_for_m_conf<bound>(pst.multiset_conf);
+    }
+
+    static constexpr inline void add_pot_proj(const sidb_cluster_ptr& c, const uint64_t sidb_ix,
+                                              const potential_projection& pp) noexcept
+    {
+        c->pot_projs[sidb_ix].add(pp);
+    }
+
+    static constexpr inline void remove_all_cluster_charge_state_occurrences(const sidb_cluster_projector_state& rm_pst,
+                                                                             const uint64_t sidb_ix) noexcept
+    {
+        rm_pst.cluster->pot_projs[sidb_ix].remove_m_conf(rm_pst.multiset_conf);
```

Comment:
I prefer ``.at()`` to access elements of a vactor.

Reply:
`remove_m_conf` is non-`const`  so `.at()`  is not possible here. I try to use `.at()` whereever possible though.